### PR TITLE
add SuiteSparse

### PR DESCRIPTION
**Pull request by NeroBurner**
optional dependency for ceres-solver as stated in Issue #855

needs LAPACK

### CODE REVIEW

File: examples/SuiteSparse/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  endif()
+endforeach()
+
+find_package(SuiteSparse CONFIG REQUIRED)
+
+# Test double library creation
+find_package(SuiteSparse CONFIG REQUIRED)
+
+# Create executable
+include(${USE_SuiteSparse})
```
Comment (ruslo) [inline examples/SuiteSparse/CMakeLists.txt]: Why this directive needed? Can it be called in `SuiteSparseConfig.cmake`?

File: examples/SuiteSparse/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+add_executable(foo foo.cpp)
+target_link_libraries(foo PRIVATE ${SuiteSparse_LIBRARIES})
```
Comment (ruslo) [inline examples/SuiteSparse/CMakeLists.txt]: Is `SuiteSparse_LIBRARIES` a list of imported targets?

Comment (NeroBurner) [inline examples/SuiteSparse/CMakeLists.txt]: yes. The template for the Use file is defined here https://github.com/jlblancoc/suitesparse-metis-for-windows/blob/master/UseSuiteSparseForInstall.cmake.in

it also calls `include_directories(${SuiteSparse_INCLUDE_DIRS})`

I think the suitesparse-metis-for-windows author was inspired by the Eigen3Config and UseEigen3 cmake files

**Comment by NeroBurner** [comment]
I've sent a Pull Request to update the suitesparse install targets
https://github.com/jlblancoc/suitesparse-metis-for-windows/pull/42

I hope it gets merged. @ruslo could you please take a look at the PR if it is sane

**Comment by NeroBurner** [comment]
`suitesparse-metis-for-windows` has the version `1.3.1`, but the official `suitesparse` version included is `4.5.1`. The included version number is also reported by the cmake-config-version file deployed by `suitesparse-metis-for-windows`

https://github.com/jlblancoc/suitesparse-metis-for-windows/releases/tag/v1.3.1

Which version should we use? I'm for `4.5.1` because it is the actual deployed version

**Comment by ruslo** [comment]
> Which version should we use? I'm for 4.5.1 because it is the actual deployed version

Okay, just leave the comment about it.

**Comment by NeroBurner** [comment]
test version using the branch hunter_v1.3.1 of https://github.com/NeroBurner/suitesparse-metis-for-windows/tree/hunter_v1.3.1

this version includes the PR https://github.com/jlblancoc/suitesparse-metis-for-windows/pull/42

File: cmake/projects/SuiteSparse/hunter.cmake [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# List of versions here...
+## suitesparse-metis-for-windows version is 1.3.1
+## suitesparse library version is 4.5.1
+hunter_add_version(
+    PACKAGE_NAME
+    SuiteSparse
+    VERSION
+    "4.5.1-p0"
+    URL
+    "https://github.com/NeroBurner/suitesparse-metis-for-windows/archive/hunter_v1.3.1.zip"
```
Comment (ruslo) [inline cmake/projects/SuiteSparse/hunter.cmake]: You can use this repository:
* https://github.com/hunter-packages/suitesparse-metis-for-windows

Comment (NeroBurner) [inline cmake/projects/SuiteSparse/hunter.cmake]: thanks, change to it tomorrow

**Comment by NeroBurner** [comment]
changed, and squashed, needs to be merged after LAPACK

**Comment by NeroBurner** [comment]
rebased on current master (which merged LAPACK)

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1927
* https://travis-ci.org/ingenue/hunter/builds/251076558?utm_source=github_status&utm_medium=notification

**Comment by NeroBurner** [comment]
passed :+1:

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1927
* https://travis-ci.org/ingenue/hunter/builds/251076558?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
> passed

I see only unit-tests passed. A lot of toolchains failed for SuiteSparse:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1927
* https://travis-ci.org/ingenue/hunter/builds/251076558?utm_source=github_status&utm_medium=notification

Feel free to exclude broken.

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1952
* https://travis-ci.org/ingenue/hunter/builds/253129606?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Uploading:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1953
* https://travis-ci.org/ingenue/hunter/builds/253136698?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.19.31

**Pull request merged**